File: application/src/main/java/com/example/server/mapper/UserMapper.java
package com.example.server.mapper;

import com.example.api.dto.UserDto;
import com.example.server.repository.entity.User;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Component
public class UserMapper {

    public User toUser(UserDto userDto) {
        if ( userDto == null ) {
            return null;
        }

        User user = new User();

        user.setName( userDto.getName() );
        user.setEmail( userDto.getEmail() );

        return user;
    }

    public User updateUser(UserDto userDto, User user) {
        if ( userDto == null ) {
            return user;
        }

        if ( userDto.getName() != null ) {
            user.setName( userDto.getName() );
        }
        if ( userDto.getEmail() != null ) {
            user.setEmail( userDto.getEmail() );
        }

        return user;
    }

    public UserDto toUserDto(User user) {
        if ( user == null ) {
            return null;
        }

        long id = 0L;
        String name = null;
        String email = null;

        if ( user.getId() != null ) {
            id = user.getId();
        }
        name = user.getName();
        email = user.getEmail();

        UserDto userDto = new UserDto( id, name, email );

        return userDto;
    }

    public List<UserDto> toUserDto(List<User> users) {
        if ( users == null ) {
            return null;
        }

        List<UserDto> list = new ArrayList<UserDto>( users.size() );
        for ( User user : users ) {
            list.add( toUserDto( user ) );
        }

        return list;
    }
}

File: application/src/main/kotlin/com/example/server/mapper/UserMapper.kt
package com.example.server.mapper

import com.example.api.model.UserDto
import com.example.server.entity.User

fun UserDto.toEntity(): User {
    return User(id = null, name = this.name!!, email = this.email!!)
}

fun User.updateEntity(userDto: UserDto): User {
    this.name = userDto.name ?: this.name
    this.email = userDto.email ?: this.email
    return this
}

fun User.toDto(): UserDto {
    return UserDto(id = this.id!!, name = this.name, email = this.email)
}

fun List<User>.toDto(): List<UserDto> {
    return map { it.toDto() }
}